### ISSUE TITLE
About code in modeling.py

### ISSUE THREAD

**Issue by AprilT0621** [uta-smile/TVT#11]
Hello!
After reading your paper and codes, I have a question about the following codes in modeling.py, can you please explain it?

        if posi_emb is not None:
            eps=1e-10
            batch_size = key_layer.size(0)
            patch = key_layer
            ad_out, loss_ad = lossZoo.adv_local(patch[:,:,1:], ad_net, is_source)
            entropy = - ad_out * torch.log2(ad_out + eps) - (1.0 - ad_out) * torch.log2(1.0 - ad_out + eps) # 标准熵H
            entropy = torch.cat((torch.ones(batch_size, self.num_attention_heads, 1).to(hidden_states.device).float(), entropy), 2) # [1, T(Kpatch)]
            trans_ability = entropy if self.vis else None   # [B*12*197]
            entropy = entropy.view(batch_size, self.num_attention_heads, 1, -1)
            attention_probs = torch.cat((attention_probs[:,:,0,:].unsqueeze(2) * entropy, attention_probs[:,:,1:,:]), 2)

In the last line why do you multiply entropy only in the 0 demension, not in all demension?

**Comment by viyjy** [comment]
Hi, thanks for your interest in our work.
The reason is that we only apply it to the CLS token. Please let me know if you have any further questions.

**Comment by AprilT0621** [comment]
Thanks to your reply. I have another question： According to the training strategy in your code, it seems that you only trained the discriminator to identify whether the input is from the source domain or target domain, but did not train the generator to fool the discriminator. Can you tell me if I get it wrong and point out where this part of codes is

**Comment by viyjy** [comment]
It is implemented by gradient reversal: https://github.com/uta-smile/TVT/blob/main/models/modeling.py#L419
Please refer to this paper: https://arxiv.org/pdf/1409.7495.pdf

**Comment by CVStack** [comment]
Is it right that domain discriminator outputs 0.5 for both source data and target data via the discriminator loss?

**Comment by viyjy** [comment]
if it is true, then it means that the discriminator cannot distinguish source data from the target data. Is that what you mean?

**Comment by CVStack** [comment]
Yeap it's right!
